### PR TITLE
Add function to get /dependency_resolvers/toolbox

### PR DESCRIPTION
**Pull request by cat-bro**
Add file for ToolDependencies API with a function `summarize_toolbox` plus a test file.

https://docs.galaxyproject.org/en/master/api/api.html#module-galaxy.webapps.galaxy.api.tool_dependencies

### CODE REVIEW

File: bioblend/_tests/TestGalaxyToolDependencies.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+
+"""
+Test functions in bioblend.galaxy.tool_dependencies
+"""
+from . import GalaxyTestBase
+
+
+class TestGalaxyToolDependencies(GalaxyTestBase.GalaxyTestBase):
+
+    def test_summarize_toolbox(self):
```
Comment (simonbray) [inline bioblend/_tests/TestGalaxyToolDependencies.py:10]: Judging by the test failures, I assume the API endpoint was not available in Galaxy versions < 20.01? If so, could you please add the `@test_util.skip_unless_galaxy('release_20.01')` decorator?

File: bioblend/galaxy/tool_dependencies/__init__.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        """
+
+        params = {}
+        if index:
+            params['index'] = str(index)
+        if tool_ids:
+            params['tool_ids'] = ','.join(tool_ids)
+        if resolver_type:
+            params['resolver_type'] = resolver_type
+        if include_containers is not None:
```
Comment (simonbray) [inline bioblend/galaxy/tool_dependencies/__init__.py:74]:
```suggestion
        if include_containers:
```
I think there is no difference between None and False here, right?

File: bioblend/galaxy/tool_dependencies/__init__.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+from bioblend.galaxy.client import Client
+
+
+class ToolDependenciesClient(Client):
+
+    def __init__(self, galaxy_instance):
+        self.module = 'dependency_resolvers'
+        super().__init__(galaxy_instance)
+
+    def summarize_toolbox(self, index=None, tool_ids=None, resolver_type=None, include_containers=None, container_type=None, index_by=None):
```
Comment (simonbray) [inline bioblend/galaxy/tool_dependencies/__init__.py:13]: I'm not sure whether these should all default to None, e.g. `include_containers` can default to requirements.

File: bioblend/galaxy/tool_dependencies/__init__.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        """
+        GET /api/dependency_resolvers/toolbox
+
+        Summarize requirements across toolbox (for Tool Management grid). This is an experiemental
```
Comment (simonbray) [inline bioblend/galaxy/tool_dependencies/__init__.py:17]:
```suggestion
        Summarize requirements across toolbox (for Tool Management grid). This is an experimental
```

File: bioblend/galaxy/tool_dependencies/__init__.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        if tool_ids:
+            params['tool_ids'] = ','.join(tool_ids)
+        if resolver_type:
+            params['resolver_type'] = resolver_type
+        if include_containers is not None:
+            params['include_containers'] = str(include_containers)
+        if container_type:
+            params['container_type'] = container_type
+        if index_by:
+            params['index_by'] = index_by
```
Comment (simonbray) [inline bioblend/galaxy/tool_dependencies/__init__.py:79]: Maybe throw an error if it's not one of tools or requirements?

File: bioblend/galaxy/tool_dependencies/__init__.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    def summarize_toolbox(self, index=None, tool_ids=None, resolver_type=None, include_containers=None, container_type=None, index_by=None):
+        """
+        GET /api/dependency_resolvers/toolbox
+
+        Summarize requirements across toolbox (for Tool Management grid). This is an experiemental
+        API particularly tied to the GUI - expect breaking changes until this notice is removed.
+
+        :type index: int
+        :param index: index of the dependency resolver
```
Comment (simonbray) [inline bioblend/galaxy/tool_dependencies/__init__.py:21]: Can you clarify a bit more what this means?

File: bioblend/galaxy/tool_dependencies/__init__.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+                          'model_class': 'NullDependency',
+                          'name': 'galaxy_sequence_utils',
+                          'version': '1.1.4'},
+                          {'cacheable': False,
+                          'dependency_type': None,
+                          'exact': True,
+                          'model_class': 'NullDependency',
+                          'name': 'bx-python',
+                          'version': '0.8.6'}],
+              'tool_ids': ['vcf_to_maf_customtrack1']}]
```
Comment (simonbray) [inline bioblend/galaxy/tool_dependencies/__init__.py:64]: Can you add a note here about the supported Galaxy versions and also that this is an admin-only feature?

**Review by simonbray — COMMENTED** [review]
Thanks @cat-bro - I'm not really familiar with this endpoint, but I wrote a few comments.

File: bioblend/galaxy/tool_dependencies/__init__.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+class ToolDependenciesClient(Client):
+
+    def __init__(self, galaxy_instance):
+        self.module = 'dependency_resolvers'
+        super().__init__(galaxy_instance)
+
+    def summarize_toolbox(self, index=None, tool_ids=None, resolver_type=None, include_containers=None, container_type=None, index_by=None):
+        """
+        GET /api/dependency_resolvers/toolbox
+
```
Comment (simonbray) [inline bioblend/galaxy/tool_dependencies/__init__.py:16]:
```suggestion
```

File: bioblend/_tests/TestGalaxyToolDependencies.py (hunk 1 above)
Comment (cat-bro) [reply bioblend/_tests/TestGalaxyToolDependencies.py:10]: Thanks!  I had not considered earlier versions

File: bioblend/galaxy/tool_dependencies/__init__.py (hunk 7 above)
Comment (cat-bro) [reply bioblend/galaxy/tool_dependencies/__init__.py:62]: Thanks @simonbray, will do.

File: bioblend/galaxy/tool_dependencies/__init__.py (hunk 6 above)
Comment (cat-bro) [reply bioblend/galaxy/tool_dependencies/__init__.py:18]: It's the index of the dependency resolver with respect to the config file.  If you use the default config, index 0 is tool shed repositories, 1 is galaxy packages, 2 is conda with the right version etc.  The output is the dependency resolution with respect to the chosen resolver. It seems easier to document this as 'index of the dependency resolver' than try to accurately explain it - I'm not sure if what I just said about it is 100% correct.

**Comment by cat-bro** [comment]
Thanks @simonbray, apologies for leaving this hanging.  I got `index_by` argument wrong, whoops.  Looks like it passed the test with `tool` instead of `tools` purely by accident.  The api will index by 'tools' for any argument other than requirements.https://github.com/galaxyproject/galaxy/blame/dev/lib/galaxy/webapps/galaxy/api/tool_dependencies.py#L224

Thank you for fixing this.

File: bioblend/galaxy/tool_dependencies/__init__.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    def __init__(self, galaxy_instance):
+        self.module = 'dependency_resolvers'
+        super().__init__(galaxy_instance)
+
+    def summarize_toolbox(self, index=None, tool_ids=None, resolver_type=None, include_containers=False, container_type=None, index_by='requirements'):
+        """
+        Summarize requirements across toolbox (for Tool Management grid).
+
+        :type index: int
+        :param index: index of the dependency resolver
```
Comment (simonbray) [inline bioblend/galaxy/tool_dependencies/__init__.py:18]:
```suggestion
        :param index: index of the dependency resolver, with respect to the
            ``container_resolvers_conf.xml`` config file.
```

**Comment by simonbray** [comment]
No worries, thanks for the PR!

**Review by simonbray — APPROVED** [review]

File: bioblend/galaxy/tool_dependencies/__init__.py (hunk 9 above)
Comment (cat-bro) [reply bioblend/galaxy/tool_dependencies/__init__.py:18]: I think this probably only works for dependency resolvers so I'll make this 'with respect to the dependency resolvers config file', will update now

**Pull request merged**